### DISCUSSION TITLE
Scripting?

### DISCUSSION THREAD

**Discussion by jumpjack** [ricktu288/ray-optics#142] [Ideas]
Is it possible to script objects?
I need to simulate a rotating and translating source of lilght.

**Comment by ricktu288** [comment]
A desmos-like variable/equation system has been proposed. See the discussion in #141.

In the current simulator, you may try to use the JavaScript console to access the global `objs` array. Call `draw()` to redraw the scene.

Updated: `objs` has been moved to `scene.objs`.

**Comment by jumpjack** [comment]
it's amazing that this tool exposes so many functions, but what about grouping them under one single collector, such as rayOptics, so developer can have a total list of them at a glance? rayOptics.add(), rayOptics.objs, ...

What I really like is that it works also locally and out of the box, please don't change this! :-)

**Reply by ricktu288** [reply]
See https://github.com/ricktu288/ray-optics?tab=readme-ov-file#use-this-project-as-a-node-module

**Reply by jumpjack** [reply]
no, not node! I hate node! I don't want to install anything, especially servers, on my machine! The coolest thing of pure JS it that it just works, without downloading anything, without installing anything, without network, without CORS issues.

**Comment by ricktu288** [comment]
I think the simplest and most straightforward way to implement this feature within in the app itself is to add an embedded JavaScript code editor. However, it seems that using `eval` or `new Function(...)` to interpret the user-provided script is generally not recommended. But I am not sure how severe it is in the current web app since there appears to be no server or user data to cause a security problem.

On the other hand, with our current availability, it seems that any other custom script proposal (e.g. the desmos-like proposal in #141) may take more than a year to implement.

**Comment by ricktu288** [comment]
See https://github.com/ricktu288/ray-optics/discussions/141#discussioncomment-9515057 for the current experimental implementation.
